packages/url_launcher/url_launcher_platform_interface/lib/link.dart

```diff
@@ -0,0 +1,99 @@
+// Copyright 2017 The Chromium Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+import 'dart:async';
+import 'dart:ui';
+
+import 'package:flutter/foundation.dart';
+import 'package:flutter/material.dart';
+import 'package:flutter/services.dart';
+
+/// Signature for a function provided by the [Link] widget that instructs it to
+/// follow the link.
+typedef FollowLink = Future<void> Function();
+
+/// Signature for a builder function passed to the [Link] widget to construct
+/// the widget tree under it.
+typedef LinkWidgetBuilder = Widget Function(
+  BuildContext context,
+  FollowLink followLink,
+);
+
+/// Signature for a delegate function to build the [Link] widget.
+typedef LinkDelegate = Widget Function(LinkInfo linkWidget);
+
+final MethodCodec _codec = const JSONMethodCodec();
+
+/// Defines where a Link URL should be open.
+///
+/// This is a class instead of an enum to allow future customizability e.g.
+/// opening a link in a specific iframe.
+class LinkTarget {
+  /// Const private constructor with a [debugLabel] to allow the creation of
+  /// multiple distinct const instances.
+  const LinkTarget._({this.debugLabel});
+
+  /// Used to distinguish multiple const instances of [LinkTarget].
+  final String debugLabel;
+
+  /// Use the default target for each platform.
+  ///
+  /// On Android, the default is [blank]. On the web, the default is [self].
+  ///
+  /// iOS, on the other hand, defaults to [self] for web URLs, and [blank] for
+  /// non-web URLs.
+  static const defaultTarget = LinkTarget._(debugLabel: 'defaultTarget');
+
+  /// On the web, this opens the link in the same tab where the flutter app is
+  /// running.
+  ///
+  /// On Android and iOS, this opens the link in a webview within the app.
+  static const self = LinkTarget._(debugLabel: 'self');
+
+  /// On the web, this opens the link in a new tab or window (depending on the
+  /// browser and user configuration).
+  ///
+  /// On Android and iOS, this opens the link in the browser or the relevant
+  /// app.
+  static const blank = LinkTarget._(debugLabel: 'blank');
+}
+
+/// Used to override the delegate that builds the link.
+LinkDelegate linkDelegate;
```

Comment:
This also looks like another platform interface method :P

Reply:
Same as my comment above, could you elaborate please? :)

Reply:
So in the federated plugins, a "PlatformInterface" is an abstract class describing the behavior that each platform must implement so the plugin works there. The Web implementation extends the PlatformInterface class and provides the web implementations, and there's normally a default MethodChannel implementation that is used by the native side.

The way you override the linkDelegate, and (especially) the builder inside LinkInfo, reminds me of the pattern above (but you reworked it from scratch), that's what I meant by "these look like PlatformInterface" classes, you're using the same pattern, but without leveraging the existing classes :P

Reply:
You're right about link delegate. There's one difference though that makes it not fit exactly well in the platform interface pattern. The default link delegate implementation has to live in `url_launcher` (the main package) because it needs to call the real `canLaunch()` and `launch()`. I'll see what I can do to use the platform interface pattern as much as possible.

OTOH, `LinkInfo` isn't being overriden anywhere. It's a class that defines the contract between `Link` and `LinkDelegate`. It's common between `url_launcher` and `url_launcher_web` that's why I put it in the platform interface package.

Reply:
About canLaunch/launch, if this ever makes it into the framework, it won't be able to use the url_launcher plugin, right? Then it'll need to be copied internally or exposed somehow. Anyway, I don't see anything blocking. (Note that there's an incoming null-safety migration to the package that might cause some merge conflicts with this PR.)